CMakeLists.txt

```diff
@@ -254,6 +244,74 @@ elseif (WITH_EEID AND OE_TRUSTZONE)
   message(FATAL_ERROR "WITH_EEID is not supported on ARM yet.")
 endif ()
 
+# Option to build and expose the libgcov such that enclave applications can use.
+# Note that this option is only effective when CODE_COVERAGE is not set.
+option(BUILD_LIBGCOV "Build the code coverage library" OFF)
+
+option(CODE_COVERAGE "Enable code coverage testing" OFF)
+if (CODE_COVERAGE)
+  if (WIN32)
+    message(
+      FATAL_ERROR
+        "The CODE_COVERAGE option currently is not supported on Windows.")
+  endif ()
+  message("-- CODE_COVERAGE set - enable code coverage tests")
+
+  if (CMAKE_C_COMPILER_ID MATCHES GNU)
+    message(FATAL_ERROR "Code coverage is currently only supported by clang.")
+  else ()
+    if (CMAKE_C_COMPILER_VERSION VERSION_LESS 7 OR CMAKE_C_COMPILER_VERSION
+                                                   VERSION_GREATER 7.99)
+      message(
+        WARNING "Code coverage may not work with clang versions other than 7.")
+    endif ()
+  endif ()
+
+  find_program(LCOV "lcov")
+  if (LCOV)
+    message("-- LCOV is found")
+  else ()
+    message(FATAL_ERROR "LCOV is not found.")
+  endif ()
+
+  add_custom_target(
```

Comment:
@mingweishih 
I referred to this PR while implementing Code Coverage for oeedger8r. https://github.com/openenclave/oeedger8r-cpp/pull/16
Some obervations:
- Moving the cmake commands to the code-coverage script simplifies cmake
- The coverage workflow seems to be to
   - Generate base info
   - Run program/tests
   - Generate coverage info
  Here we don't seem to be running the program/tests between the baseline info and coverage info.
  Can you double check what is the correct order of steps?
- GCC docs say that -coverage or -fprofile-arcs must also be passed to the linker.

Reply:
My original thought is simply that we have `make test` or `ctest` to run the test already, so I `make code-coverage` as complementary to that.

Note that even generating base info (with `lcov -i`) can be independent from the test run.

Reply:
But the lcov docs don't recommend running `lcov -i` after the test:
https://linux.die.net/man/1/lcov

**--initial**
Capture initial zero coverage data.
Run lcov with -c and this option on the directories containing .bb, .bbg or .gcno files **before running any test case.** The result is a "baseline" coverage data file that contains zero coverage for every instrumented line.  **Combine this data file (using lcov -a) with coverage data files captured after a test run to ensure that the percentage of total lines covered is correct even when not all source code files were loaded during the test.**

    Recommended procedure when capturing data for a test case:

    1. create baseline coverage data file

        # lcov -c -i -d appdir -o app_base.info

    2. perform test

        # appdir/test

    3. create test coverage data file

        # lcov -c -d appdir -o app_test.info

    4. combine baseline and test coverage data

        # lcov -a app_base.info -a app_test.info -o app_total.info

Reply:
Sure, I'll take the recommendation then.

Reply:
I don't think it is a must have for this PR. You can investigate and make changes in next PR.

Reply:
That also sounds good to me! I'm waiting for @BRMcLaren to unblock this PR from the CI/CI infrastructure so that we can merge it.

I'll create an issue/PR for the follow-up.